frontend/js/src/artist-similarity/src/Data.tsx

```diff
@@ -0,0 +1,167 @@
+import Input from "./Input";
+import tinycolor from "tinycolor2";
+
+type ArtistType = {
+    artist_mbid: string;
+    name: string;
+    comment: string;
+    type: string;
+    gender: string;
+    score?: number;
+    reference_mbid?: string;
+}
+    
+type MarkupResponseType = {
+    data: string;
+    type: "markup";
+}
+
+type DatasetResponseType = {
+    columns: Array<string>;
+    data: Array<ArtistType>;
+    type: "dataset";
+}
+
+type ApiResponseType = MarkupResponseType | DatasetResponseType;
+
+type NodeType = {
+    id: string;
+    artist_mbid: string;
+    size: number;
+    color: string;
+    score: number;
+}
+
+type LinkType = {
+    source: string;
+    target: string;
+    distance: number;
+}
+
+type GraphDataType = {
+    nodes: Array<NodeType>;
+    links: Array<LinkType>;
+}
+
+const colorGenerator = ():  tinycolor.Instance=> {
+    var color = tinycolor("hsv(" + Math.random() * 360 + ", 100%, 90%)");
+    return color;
+}
+
+/*const mixColor = (color1: ColorType, color2: ColorType, weight: number): ColorType => {
+    var red = Math.floor((color2.red - color1.red) * weight + color1.red);
+    var green = Math.floor((color2.green - color1.green) * weight + color1.green);
+    var blue = Math.floor((color2.blue - color1.blue) * weight + color1.blue);          
+    var opacity = Math.floor((color2.opacity - color1.opacity) * weight + color1.opacity);
+
+    return Color(red, green, blue);
+}*/
+
+const Data = () => {
+    const ARTIST_MBID = "8f6bd1e4-fbe1-4f50-aa9b-94c450ec0f11";
+    const SIMILAR_ARTISTS_LIMIT_VALUE = 18;
+    const BASE_URL = "https://labs.api.listenbrainz.org/similar-artists/json?algorithm=session_based_days_7500_session_300_contribution_5_threshold_10_limit_100_filter_True_skip_30&artist_mbid=";
+    const LINK_DIST_MULTIPLIER = 200;
+    const MIN_LINK_DIST = 50;
+    const MAIN_NODE_SIZE = 150;
+    const SIMILAR_NODE_SIZE = 85;
+    const BACKGROUND_ALPHA = 0.2;
+    const COLOR_MIX_WEIGHT = 0.3;
+    
+    var color1 = colorGenerator();
+    var color2 = colorGenerator();
+
+    const [similarArtists, setSimilarArtists] = useState<DatasetResponseType>();
+    const [artist, setArtist] = useState<DatasetResponseType>();
+    const [similarArtistsLimit, setSimilarArtistsLimit] = useState(SIMILAR_ARTISTS_LIMIT_VALUE);
+    const [colors, setColors] = useState([color1, color2]);
+    const [artistMBID, setArtistMBID] = useState(ARTIST_MBID);
+
+    var scoreList: Array<number> = [];
+    
+    const fetchData = async (ARTIST_MBID: string): Promise<void> => {
+        try {
+          const response = await fetch(BASE_URL + ARTIST_MBID);
+          const data = await response.json();
+          console.log(data);
+          processData(data);
+        }
+        catch (error){
+          //Error message goes here.
+          alert("Something went wrong while loading information, please try again");
+        }
+    }
+    
+    const processData = (data: Array<ApiResponseType>): void => {
```

Comment:
I see the type of `data` in this function is defined as `Array<ApiResponseType>`, unless I'm missing something I think that might be incorrect. Should it not be `ApiResponseType` instead, which itself contains a `data` property?

I would recommend to choose more descriptive names to avoid confusions like this, maybe: `similarArtistResponse: ApiResponseType`

After you change this type I think you'll find some of the code below will need changing, something like
```
if(similarArtistResponse.type === "dataset"){
// deconstructive assignment
  const [mainArtist, ...similarArtists] = similarArtistResponse.data;
setArtist(mainArtist);
setSimilarArtists(similarArtists);
}
```
For deconstructive assignment, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

Reply:
Actually API response contains is an array. It either contains elements which are markup response or dataset response. The data attribute exists only on the dataset responses and not directly on ApiResponseType. I have modified ApiResponseType to be an array itself which I believe is much clearer.